--- source:thread_state
+++ verus:thread_state
@@ -1,9 +1,13 @@
-    pub fn thread_state(&self) -> &ThreadState {
+    pub fn thread_state(&self) -> (result: &ThreadState)
+        ensures
+            result.spec_id() == self.spec_id(),
+            result@ == self.spec_state()@,
+    {
         match self {
-            ThreadRef::Ready(thread) => thread.thread_state(),
-            ThreadRef::Running(thread) => thread.thread_state(),
-            ThreadRef::Sleeping(thread) => thread.thread_state(),
-            ThreadRef::Interrupted(thread) => thread.thread_state(),
-            ThreadRef::Zombie(thread) => thread.thread_state(),
+            ThreadRefMutModel::Ready(state) => state,
+            ThreadRefMutModel::Running(state) => state,
+            ThreadRefMutModel::Sleeping(state) => state,
+            ThreadRefMutModel::Interrupted(state) => state,
+            ThreadRefMutModel::Zombie(state) => state,
         }
     }
