--- source:index_unchecked
+++ verus:index_unchecked
@@ -1,5 +1,14 @@
-    fn index_unchecked(&self, index: usize) -> (usize, usize) {
-        let word: usize = index / u8::BITS as usize;
-        let bit: usize = index % u8::BITS as usize;
+    fn index_unchecked(&self, bit_index: usize) -> (result: (usize, usize))
+        requires
+            self.inv(),
+            bit_index < self.number_of_bits,
+        ensures
+            result.0 < self.bits@.len(),
+            result.1 < u8::BITS as usize,
+            result.0 as int == bit_index as int / (u8::BITS as int),
+            result.1 as int == bit_index as int % (u8::BITS as int),
+    {
+        let word: usize = bit_index / u8::BITS as usize;
+        let bit: usize = bit_index % u8::BITS as usize;
         (word, bit)
     }
