--- source:map_kpage
+++ verus:map_kpage
@@ -1,69 +1,16 @@
-    pub fn map_kpage<T: Fn() -> Result<PageTable<PageTableStorage>, Error>>(
+    pub fn map_kpage(
         &mut self,
-        kpage: KernelPage,
-        vaddr: PageAligned<VirtualAddress>,
-        page_table_allocator: T,
-    ) -> Result<(), Error> {
-        let pt_vaddr: PageTableAddress = PageTableAddress::new(PageTableAligned::from_raw_value(
-            ::sys::mm::align_down(vaddr.into_raw_value(), PGTAB_ALIGNMENT),
-        )?);
-
-        // Get the corresponding page directory entry.
-        let pde: PageDirectoryEntry = match self.pgdir.read_pde(pt_vaddr) {
-            Some(pde) => pde,
-            None => {
-                let reason: &str = "failed to read page directory entry";
-                error!("{reason}");
-                return Err(Error::new(ErrorCode::TryAgain, reason));
-            },
-        };
-
-        // Check if page table does not exist.
-        if !pde.is_present() {
-            let page_table: PageTable<PageTableStorage> = page_table_allocator()?;
-
-            // FIXME: do not be so open about permissions.
-            self.pgdir.map(
-                pt_vaddr,
-                page_table.physical_address()?,
-                false,
-                AccessPermission::RDWR,
-            )?;
-
-            //===================================================================
-            // NOTE: if we fail beyond this point we should unmap the page table.
-            //===================================================================
-
-            // Add page table to the list of kernel page tables.
-            self.kernel_page_tables
-                .push_back(Rc::new(RefCell::new((pt_vaddr, page_table))));
-        };
-
-        // Get corresponding page table.
-        for entry in self.kernel_page_tables.iter_mut() {
-            if entry.borrow().0.into_raw_value() == pt_vaddr.into_raw_value() {
-                // Map the page to the target virtual address space.
-                // FIXME: do not be so open about permissions and caching.
-                entry.borrow_mut().1.map(
-                    PageAddress::new(vaddr),
-                    kpage.frame_address(),
-                    true,
-                    true,
-                    false,
-                    AccessPermission::RDWR,
-                )?;
-
-                // Add the kernel page to the list of kernel pages.
-                self.kernel_pages.push_back(Rc::new(RefCell::new(kpage)));
-
-                // Reload page directory to force a TLB flush.
-                self.load()?;
-
-                return Ok(());
-            }
-        }
-
-        let reason: &str = "page table not found";
-        error!("{reason}");
-        Err(Error::new(ErrorCode::NoSuchEntry, reason))
+        frame_addr: FrameAddress,
+        vaddr: usize,
+        access: AccessPermission,
+    ) -> (result: Result<(), Error>)
+        requires
+            old(self).inv(),
+            spec_is_kernel_addr(vaddr as int),
+            vaddr as int % PAGE_SIZE as int == 0,
+        ensures
+            self.inv(),
+            self@.mapping_count == old(self)@.mapping_count,
+    {
+        unimplemented!()
     }
