--- source:find_thread
+++ verus:find_thread
@@ -1,26 +1,11 @@
-    pub fn find_thread(&self, tid: ThreadIdentifier) -> Option<ThreadRef<'_>> {
-        // Search in the list of interrupted threads.
-        if let Some(thread) = self
-            .interrupted_threads
-            .iter()
-            .find(|thread| thread.id() == tid)
-        {
-            return Some(ThreadRef::Interrupted(thread));
+    pub fn find_thread(&self, tid: u64) -> (result: Ghost<Option<int>>)
+        requires
+            self.wf(),
+        ensures
+            result@ == self@.spec_find_thread(tid as int),
+    {
+        proof {
+            Self::lemma_find_thread_view_equiv(self, tid);
         }
-
-        // Search in the list of sleeping threads.
-        if let Some(sleeping_threads) = &self.sleeping_threads {
-            if let Some(thread) = sleeping_threads.iter().find(|thread| thread.id() == tid) {
-                return Some(ThreadRef::Sleeping(thread));
-            }
-        }
-
-        // Search in the list of zombie threads.
-        if let Some(zombie_threads) = &self.zombie_threads {
-            if let Some(thread) = zombie_threads.iter().find(|thread| thread.id() == tid) {
-                return Some(ThreadRef::Zombie(thread));
-            }
-        }
-
-        None
+        Ghost(self.spec_find_thread(tid))
     }
