--- source:index_unchecked
+++ verus:index_unchecked
@@ -1,5 +1,6 @@
-    fn index_unchecked(&self, index: usize) -> (usize, usize) {
-        let word: usize = index / u8::BITS as usize;
-        let bit: usize = index % u8::BITS as usize;
+    fn index_unchecked(&self, bit_index: usize) -> (usize, usize)
+    {
+        let word: usize = bit_index / u8::BITS as usize;
+        let bit: usize = bit_index % u8::BITS as usize;
         (word, bit)
     }
