--- source:down
+++ verus:down
@@ -1,19 +1,13 @@
-    pub unsafe fn down(&self) -> Result<(), SleepError> {
-        loop {
-            if self
-                .value
-                .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |value| {
-                    if value == 0 {
-                        None
-                    } else {
-                        Some(value - 1)
-                    }
-                })
-                .is_ok()
-            {
-                return Ok(());
-            }
-
-            self.sleeping.wait(None)?;
-        }
+    pub fn down(&mut self, ctx: Ghost<CallerContext>)
+        requires
+            old(self).wf(),
+            old(self).spec_is_available(),
+            ctx@.safe_for_down(),
+        ensures
+            self@.value == old(self)@.value - 1,
+            self@.waiters == old(self)@.waiters,
+            self.wf(),
+    {
+        proof { reveal(Semaphore::wf); }
+        self.value = self.value - 1;
     }
