--- source:try_down
+++ verus:try_down
@@ -1,17 +1,19 @@
-    pub fn try_down(&self) -> Result<(), Error> {
-        if self
-            .value
-            .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |value| {
-                if value == 0 {
-                    None
-                } else {
-                    Some(value - 1)
-                }
-            })
-            .is_ok()
-        {
-            return Ok(());
+    pub fn try_down(&mut self) -> (result: bool)
+        requires
+            old(self).wf(),
+        ensures
+            result == old(self).spec_is_available(),
+            result ==> self@.value == old(self)@.value - 1,
+            !result ==> self@ == old(self)@,
+            !result ==> self.spec_is_exhausted(),
+            self@.waiters == old(self)@.waiters,
+            self.wf(),
+    {
+        proof { reveal(Semaphore::wf); }
+        if self.value > 0 {
+            self.value = self.value - 1;
+            true
+        } else {
+            false
         }
-
-        Err(Error::new(ErrorCode::TryAgain, "semaphore is busy"))
     }
