--- source:top
+++ verus:top
@@ -1,3 +1,9 @@
-    pub fn top(&self) -> PageAligned<VirtualAddress> {
-        PageAligned::from_raw_value(self.base.into_raw_value() + self.size()).unwrap()
+    pub fn top(&self) -> (result: PageAlignedAddr)
+        requires
+            self.inv(),
+        ensures
+            result.inv(),
+            result.spec_addr() == self.spec_top(),
+    {
+        PageAlignedAddr::from_raw_unchecked(self.base_addr + USER_STACK_SIZE)
     }
